app/grandchallenge/discussion_forums/templates/discussion_forums/forumpost_list.html

```diff
@@ -1,46 +1,92 @@
-            <a href="{% url 'discussion-forums:topic-delete' slug=topic.slug %}" class="btn btn-danger btn-sm">
-                <i class="fas fa-trash mr-1"></i>Delete topic
-            </a>
-        {% endif %}
-    </div>
-    {# TODO: lock topic button #}
+    {% if topic %}
+        <div class="h3 my-3" title="{% if topic.is_locked %}This topic is locked.{% endif %}">
+            {% if topic.is_locked %}<i class="fas fa-lock mr-2 fa-xs badge-dark p-2 rounded"></i>{% endif %}
+            {{ topic.subject }}
+        </div>
+    {% else %}
+        <h3>My posts</h3>
+    {% endif %}
+
+    {% if topic %}
+        {% get_obj_perms request.user for topic as "user_perms" %}
+        <div class="row">
+            <div class="col-8">
+                {% if 'create_topic_post' in user_perms %}
+                    <div class="d-inline" title="{% if topic.is_locked %}This topic is locked. You cannot reply to it.{% endif %}">
+                        <a hx-get="{% url 'discussion-forums:post-create' slug=topic.slug %}" hx-swap="beforeend show:bottom" hx-target="#post-list" class="btn btn-primary btn-sm {% if topic.is_locked %}disabled{% endif %}">
+                            <i class="fas fa-comment mr-1"></i> Reply
+                        </a>
+                    </div>
+                {% endif %}
+                {% if 'delete_forumtopic' in  user_perms %}
+                    <a href="{% url 'discussion-forums:topic-delete' slug=topic.slug %}" class="btn btn-danger btn-sm">
+                        <i class="fas fa-trash mr-1"></i>Delete topic
+                    </a>
+                {% endif %}
+                {% if 'change_forumtopic' in user_perms and not topic.is_locked %}
+                    <form class="d-inline"
+                          hx-post="{% url 'discussion-forums:topic-lock-update' slug=topic.slug %}"
+                          hx-vals='{"is_locked": "True"}'
+                          hx-swap="outerHTML"
+                          hx-target="body"
+                    >
+                        {% csrf_token %}
+                        <button type="submit" class="btn btn-warning btn-sm btn-subscription"><i class="fas fa-lock mr-1"></i>
+                            Lock topic
+                        </button>
+                    </form>
+                {% elif 'change_forumtopic' in user_perms and topic.is_locked %}
+                    <form class="d-inline"
+                          hx-post="{% url 'discussion-forums:topic-lock-update' slug=topic.slug %}"
+                          hx-vals='{"is_locked": "False"}'
+                          hx-swap="outerHTML"
+                          hx-target="body"
+                    >
+                        {% csrf_token %}
+                        <button type="submit" class="btn btn-warning btn-sm btn-subscription"><i class="fas fa-lock-open mr-1"></i>
+                            Unlock topic
+                        </button>
+                    </form>
```

Comment:
FYI - htmx is not needed here if you are doing target="body". This can be a normal Django form view and post which will then redirect. Then you shouldn't need the `topic_type_matching_dict` or any of the special handlers for htmx.

Reply:
True. The `topic_type_matching_dictionary` is still necessary though. That's for the mirgation.

Reply:
Ah I see!